### ISSUE TITLE
Missing PEP/PubSub feature announcement

### ISSUE THREAD

**Issue by tmolitor-stud-tu** [maranda/metronome#549]
The XEP-0030 disco on the user's account does not announce `http://jabber.org/protocol/pubsub#subscribe` support (despite announcing `http://jabber.org/protocol/pubsub#publish` and others).

This is likely a bug and prevents Monal from detecting PEP (XEP-0163) as supported which in turn is needed for OMEMO to work properly.

**Comment by Neustradamus** [comment]
@tmolitor-stud-tu: To follow this ticket.

**Comment by Letterus** [comment]
Looks like we only need to add "subscribe" over there: https://github.com/maranda/metronome/blob/master/plugins/mod_pubsub.lua#L105 ?

**Comment by Letterus** [comment]
Hm, no. There is "subscribe", but I don't get on what it announcement depends.

**Comment by maranda** [comment]
No it's not, https://github.com/maranda/metronome/blob/master/plugins/auxlibs/pep.lib.lua#L201 it's again just a question of implementation, if you set the access model to open to the public keys node that won't be an issue also Metronome MUC does support routing of arbitrary payloads directly to the bare jid and pep is added by default.

**Comment by tmolitor-stud-tu** [comment]
I'm not talking about the public keys node but about the devicelist node.

How can I (as a client) get updates of the devicelist if I can't subscribe to the devicelist?
Sure, I can proactively fetch every devicelist after every resource bind (that would create more traffic than needed, but is doable), but what about updates of the devicelist (new added devices or removed devices)? How do these updates on a remote account reach my device?

This is security relevant: if I still use the old devicelist I might encrypt for a device already removed.
And UX: I might not encrypt for a new device already added.

As for the MUC routing stuff: querying pep through the MUC is something not usually done by any client. But in theory it would allow omemo encrypted public channels, which is kind of nice but on the other hand useless because public channels are...ehm...public, so encryption is kinda useless.

**Comment by maranda** [comment]
Are we talking about OMEMO in groupchats or something else..?
Anyways XEP-0384 itself recommends that access model of the devicelist is set to open to allow establishing of sessions by entities without presence subscription in 5.3.1 and in 5.8.2 it says:
> Before sending a message a participant MUST explicitly fetch device lists (if not already cached) for each of the members.

So at regular intervals you should explicitly refetch the device list anyways, I agree it's not exactly the best, but neither is relying exclusively on node update broadcasts because:
 A) a server may arbitrarily decide to not send you an update (if it's resource constrained for example)
 B) stanzas may get lost in transition if there're connectivity issues between the servers
 C) and that's this case, the server may have a limited PubSub feature set available to PEP and not support explicit subscriptions

 7.1 also clearly states what are the PEP requirements for OMEMO support server-side and #subscribe is not one of 'em

**Comment by tmolitor-stud-tu** [comment]
Yes, we are talking about omemo in groupchats.

No, fetching the devicelist every time before sending a message is not feasible. And caching is a hack because security, UX as well as the generated traffic greatly depend on the selected caching timeout and usage scenario (for some scenarios a longer timeout may be okay, for others it won't).
Especially on mobile devices!

- As of case A: to my knowledge neither ejabberd, nor prosody or openfire arbitrarily filter out pep stanzas, so I consider that a badly behaving server.
- Case B on the other hand should not occur very often and a broken s2s will introduce other problems, too (messages not being delivered, for example). I consider that out of scope (and the proper fix would be to use something like XEP-0198 on s2s connections or invent some other reliability improving measure).
- Yes C is your server.

> 7.1 also clearly states what are the PEP requirements for OMEMO support server-side and #subscribe is not one of 'em

yes, those are the absolute minimum requirements to make it work _somehow_. In case of missing subscribe this means regular polling. I won't resort to such hacks given that other popular servers (ejabberd, prosody, openfire) all support #subscribe (and polling is especially bad on mobile devices not being permanently connected).

**To sum it up:** Yes, it is possible to work around missing #subscribe support, but these hacks are far from optimal and I therefore consider a server not supporting #subscribe as not being ready for omemo.

Sidenote: the omemo spec implemented by most clients is OMEMO1 specced in XEP version 0.3.0 over here: https://xmpp.org/extensions/attic/xep-0384-0.3.0.html

**Comment by maranda** [comment]
Yes so to sum it up, Monal problem, Conversation works case closed :)
